### PR TITLE
Introduce NonterminalKind for more type-safe mbe parsing

### PR DESCRIPTION
**Pull request by matklad**
It encapsulate the (part of) the interface between the parser and
macro by example (macro_rules) parser.

The second bit is somewhat more general `parse_ast_fragment`, which is
the reason why we keep some `parse_xxx` functions as public.

### CODE REVIEW

**Comment by matklad** [comment]
r? @petrochenkov 

I also want to change `MetaVarDecl(Span, Ident /* name to bind */, Ident /* kind of nonterminal */),` to `MetaVarDecl(Span, Ident /* name to bind */, Option<MbeFragment>),`

**Comment by matklad** [comment]
>I also want to change MetaVarDecl(Span, Ident /* name to bind */, Ident /* kind of nonterminal */), to MetaVarDecl(Span, Ident /* name to bind */, Option<MbeFragment>),

Ah, `MetaVarDecl` is also serializable. I think it makes sense to keep it as is then and not to mix serialization into parser's interface.

**Comment by mark-i-m** [comment]
This code is quite out of cache for me, but any improvements are very well! 🎉

File: src/librustc_parse/mbe_fragment.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
+use rustc_ast::ptr::P;
+use rustc_ast::token::{self, Nonterminal, Token};
+use rustc_ast_pretty::pprust;
+use rustc_errors::PResult;
+use rustc_span::symbol::{kw, sym, Ident, Symbol};
+
+use crate::parser::{FollowedByType, Parser, PathStyle};
+
+#[derive(Copy, Clone)]
+pub enum MbeFragment {
```
Comment (petrochenkov) [inline src/librustc_parse/mbe_fragment.rs:10]: FWIW, we already have this enum in `token.rs`, it's called `enum Nonterminal`, except that it has data in addition to tags.
Perhaps these two enums should live together in `rustc_ast` as `Nonterminal` and `NonterminalKind`.

The `may_begin_with` is also very similar to `*can_begin*` functions in `token.rs`.

Moving the nonterminal parsing to `rustc_parse` is certainly a good move.

Comment (petrochenkov) [reply src/librustc_parse/mbe_fragment.rs:10]: Perhaps it's also a good opportunity to try [`#![feature(arbitrary_enum_discriminant)]`](https://github.com/rust-lang/rfcs/blob/master/text/2363-arbitrary-enum-discriminant.md).

**Comment by matklad** [comment]
Switched to `NonterminalKind`! This actually I think made it possible to use it in the `MetaVarDecl`, but that would be a somewhat bigger refactoring with higher risk of breaking something, so it might make sense to review the two separately.

**Comment by matklad** [comment]
Yup, so the second commit uses NonterminalKind for MetaVarDecl as well, ready for review now!

**Comment by matklad** [comment]
The remaining weirdness with Option is due to https://github.com/rust-lang/rust/issues/40107 I belive. I'll tackle that next, seems fine to try to hard error now.

File: src/librustc_parse/lib.rs [hunk 2]
```diff
@@ -26,6 +26,7 @@ pub const MACRO_ARGUMENTS: Option<&'static str> = Some("macro arguments");
 pub mod parser;
 use parser::{emit_unclosed_delims, make_unclosed_delims_error, Parser};
 pub mod lexer;
+pub mod nonterminal;
```
Comment (petrochenkov) [inline src/librustc_parse/lib.rs:29]: `librustc_parse/nonterminal.rs` -> `librustc_parse/parser/nonterminal.rs`

File: src/librustc_parse/parser/mod.rs [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
-pub use path::PathStyle;
-mod diagnostics;
-mod generics;
 mod stmt;
-use diagnostics::Error;
+mod ty;
 
 use crate::lexer::UnmatchedBrace;
+use diagnostics::Error;
+pub use path::PathStyle;
```
Comment (matklad) [inline src/librustc_parse/parser/mod.rs:14]: Are there any word of mouth and/or written guidelines about the ordering of mods and uses? I've fixed this negatively, by not mixing uses and modules in random order, but i am wondering what the positive fix would look like. In rust-analyer, we are somewhat picky about this: https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#order-of-imports

Comment (petrochenkov) [reply src/librustc_parse/parser/mod.rs:14]: I don't think there's an official guideline, but the most common order is
```
use crate::things;

use other_crate::things;

use std::things;

mod foo;

/* Other items */
```

**Comment by matklad** [comment]
Seems to be a spurious network error:

```
    Updating crates.io index
error: failed to get `cc` as a dependency of package `bootstrap v0.0.0 (D:\a\rust\rust\src\bootstrap)`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  error inflating zlib stream; class=Zlib (5)
failed to run: D:\a\rust\rust\build\i686-pc-windows-msvc\stage0\bin\cargo.exe build --manifest-path D:\a\rust\rust\src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:00:16
make: *** [Makefile:60: prepare] Error 1
Command failed. Attempt 5/5:
    Updating crates.io index
error: failed to get `cc` as a dependency of package `bootstrap v0.0.0 (D:\a\rust\rust\src\bootstrap)`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  error inflating zlib stream; class=Zlib (5)
```

@bors retry

**Pull request merged**